cradle-core/src/main/java/com/exactpro/cradle/messages/StoredMessageBatch.java

```diff
@@ -166,24 +162,32 @@ public StoredMessage addMessage(MessageToStore message) throws CradleStorageExce
 			StoredMessage lastMsg = getLastMessage();
 			if (message.getIndex() > 0)  //I.e. message index is set
 			{
-				messageIndex = message.getIndex();
+				long messageIndex = message.getIndex();
 				if (messageIndex <= lastMsg.getIndex())
 					throw new CradleStorageException("Message index should be greater than "+lastMsg.getIndex()+" for the batch to contain sequenced messages, but in your message it is "+messageIndex);
 				if (messageIndex != lastMsg.getIndex()+1)
 					logger.debug("Message index should be "+(lastMsg.getIndex()+1)+" for the batch to contain strictly sequenced messages, but in your message it is "+messageIndex);
 			}
-			else
-				messageIndex = lastMsg.getIndex()+1;
+			if (lastMsg.getTimestamp().isAfter(message.getTimestamp()))
+				throw new CradleStorageException(
+						"Message timestamp should be greater than last message timestamp in batch '" + lastMsg.getTimestamp()
+								+ "' but in your message it is '" + message.getTimestamp() + "'");
 		}
-		
+
+		return addMessageInternal(message);
+	}
+
+	protected StoredMessage addMessageInternal(MessageToStore message)
+	{
+		long messageIndex = message.getIndex() >= 0 ? message.getIndex() : getLastMessage().getIndex()+1;
```

Comment:
`getLastMessage()` can return `null` here.

Reply:
I think it is impossible. There are two cases of calling the addMessageInternal() method:
1. Calling from addMessage() method. All needed fields of first message are already checked in addMessage(). When next message hasn't index, first message is there already in messages collection.
2. Calling from DetailedMessageBatchEntity. All messages from db have index.

Reply:
But for possible future use, I may add checking for the last message and throwing an exception

Reply:
I decided not to add the sequence check of the first message in the `addMessageInternal()` method